### PR TITLE
Add batch BHS contract

### PR DESCRIPTION
**Pull request by makramkd**
Add a batch blockhash store contract to store blockhashes with much higher throughput.

In a single transaction we could store up to a 100 blockhashes, instead of needing 100 transactions to do so with the vanilla blockhash store contract. This is invaluable on networks like Matic.

### CODE REVIEW

File: contracts/hardhat.config.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
-import '@typechain/hardhat'
 import '@nomiclabs/hardhat-ethers'
 import '@nomiclabs/hardhat-etherscan'
 import '@nomiclabs/hardhat-waffle'
-import 'hardhat-contract-sizer'
+import '@typechain/hardhat'
```
Comment (Nic0S) [inline contracts/hardhat.config.ts:4]: Why these changes?

File: contracts/src/v0.8/dev/BatchBHS.sol [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+pragma solidity 0.8.6;
+
+/**
+ * @title BatchBHS
+ * @notice The BatchBHS contract acts as a proxy to write many blockhashes to the
+ *   provided BlockhashStore contract efficiently in a single transaction. This results
+ *   in plenty of gas savings and higher throughput of blockhash storage, which is desirable
+ *   in times of high network congestion.
+ */
+contract BatchBHS {
```
Comment (Nic0S) [inline contracts/src/v0.8/dev/BatchBHS.sol:11]: Should this be called BatchBlockhashStore just to match the original BlockhashStore? The name doesn't seem to long to me

File: contracts/src/v0.8/dev/BatchBHS.sol [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  constructor(address blockhashStoreAddr) {
+    BHS = BlockhashStore(blockhashStoreAddr);
+  }
+
+  /**
+   * @notice stores blockhashes of the given block numbers in the configured blockhash store, assuming
+   *   they are availble though the blockhash() instruction.
+   * @param blockNumbers the block numbers to store the blockhashes of. Must be available via the
+   *   blockhash() instruction, otherwise this function call will revert.
```
Comment (Nic0S) [inline contracts/src/v0.8/dev/BatchBHS.sol:22]: Should the whole thing revert, or would it make sense to just skip over any that aren't available? Makes it a little safer in case only one is past the 256 block limit

File: contracts/src/v0.8/dev/BatchBHS.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    for (uint256 i = 0; i < blockNumbers.length; i++) {
+      BHS.storeVerifyHeader(blockNumbers[i], headers[i]);
+    }
+  }
+
+  /**
+   * @notice retrieves blockhashes of all the given block numbers from the blockhash store, if available.
+   * @param blockNumbers array of block numbers to fetch blockhashes for
+   */
+  function getBlockhashes(uint256[] memory blockNumbers) external view returns (bytes32[] memory) {
```
Comment (Nic0S) [inline contracts/src/v0.8/dev/BatchBHS.sol:46]: Similar to the other comment, I think this one may not be that useful if it reverts if any are missing. The value would be knowing which ones are present / missing

**Review by Nic0S — COMMENTED** [review]
nice, looks great and good documentation. A couple comments mainly on partial reverts in BatchBHS

File: contracts/hardhat.config.ts (hunk 1 above)
Comment (makramkd) [reply contracts/hardhat.config.ts:4]: I think these were just import ordering changes by the formatter to order them alphabetically.

File: contracts/src/v0.8/dev/BatchBHS.sol (hunk 3 above)
Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol:22]: I think if we want to make it so that it doesn't revert, would we want to return the block numbers that failed to store? Maybe something like:

```solidity
// returns block numbers that we couldn't store
function store(uint256[] memory blockNumbers) public returns (uint256[]);
```

We could also silently skip over them, though in that case it might be harder to debug. E.g if one block number is too far back and we return that, the caller could call `storeVerifyHeader` with the farthest back block that we were able to store with this method.

Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol:22]: The advantage of having this revert right away is code and usage simplicity, I guess.

File: contracts/src/v0.8/dev/BatchBHS.sol (hunk 4 above)
Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol:46]: Yeah, actually I think you're right. Seems like we have a couple of options on how to do that here. 

We could simply have the entry in the returned array as `0x0` if the blockhash is missing (this is what the `blockhash` instruction returns if the block is too old), and the caller would have to match the index of the 0x0 in the returned array to the input array to get the block number that was missing.

We could additionally return a separate array with the block numbers that don't have an associated blockhash in the BHS, though that might be confusing as a function interface.

Comment (Nic0S) [reply contracts/src/v0.8/dev/BatchBHS.sol:46]: In practice I feel like we only ever need to know which are missing / present, we could probably just return a list of block numbers for which the blockhashes aren't present?

If only BlockhashStore's s_blockhashes wasn't internal..would make life a lot easier haha

File: contracts/src/v0.8/dev/BatchBHS.sol (hunk 2 above)
Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol]: Done.

File: contracts/hardhat.config.ts (hunk 1 above)
Comment (makramkd) [reply contracts/hardhat.config.ts:4]: Agreed!

File: contracts/src/v0.8/dev/BatchBHS.sol (hunk 3 above)
Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol]: Done.

File: contracts/src/v0.8/dev/BatchBHS.sol (hunk 4 above)
Comment (makramkd) [reply contracts/src/v0.8/dev/BatchBHS.sol]: Opted to return `0x0` in the returned array to indicate missing blockhash for the associated block number at that index.

**Review by Nic0S — APPROVED** [review]
awesome

**Review by Nic0S — APPROVED** [review]
